Userland/Libraries/LibGfx/Filters/GrayscaleFilter.h

```diff
@@ -0,0 +1,45 @@
+/*
+ * Copyright (c) 2021, David Savary <david.savarymartinez@gmail.com>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#pragma once
+
+#include <LibGfx/Filters/Filter.h>
+
+namespace Gfx {
+
+class GrayscaleFilter : public Filter {
+public:
+    GrayscaleFilter() { }
+    virtual ~GrayscaleFilter() { }
+
+    virtual char const* class_name() const override { return "GrayscaleFilter"; }
+
+    virtual void apply(Bitmap& target_bitmap, IntRect const& target_rect, Bitmap const& source_bitmap, IntRect const& source_rect) override
+    {
+        // source_rect should be describing the pixels that can be accessed
+        // to apply this filter, while target_rect should describe the area
+        // where to apply the filter on.
+        VERIFY(source_rect.size() == target_rect.size());
+        VERIFY(target_bitmap.rect().contains(target_rect));
+        VERIFY(source_bitmap.rect().contains(source_rect));
+
+        for (auto i = 0; i < source_rect.width(); ++i) {
+            ssize_t source_i = i + source_rect.x();
+            ssize_t target_i = i + target_rect.x();
+            for (auto j = 0; j < source_rect.height(); ++j) {
+                ssize_t source_j = j + source_rect.y();
+                ssize_t target_j = j + target_rect.y();
```

Comment:
This seems to be going column-by-column rather than row-by-row. Also, `i` and `j` seem to make less contextual sense than `x` and `y` (which are mentioned by the rects).

Reply:
This makes total sense to me. Maybe we should also change the other filters so that they also access the image row wise.